repose-aggregator/components/services/http-client-service/http-client-service-api/src/main/java/org/openrepose/core/services/httpclient/CachingHttpClientContext.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * _=_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=
+ * Repose
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Copyright (C) 2010 - 2015 Rackspace US, Inc.
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+package org.openrepose.core.services.httpclient;
+
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.protocol.HttpContext;
+
+/**
+ * Adaptor class that provides convenient, type-safe getters and setters
+ * for cache control settings.
+ */
+public class CachingHttpClientContext extends HttpClientContext {
+
+    /**
+     * Designates whether or not the cache should be used.
+     * <p>
+     * If used, the {@link org.apache.http.client.HttpClient} will attempt
+     * to satisfy the HTTP request associated with this context with a
+     * previously cached interaction.
+     */
+    public static final String CACHE_USE = "repose.cache.use";
+
+    /**
+     * Identifies what cache data may be used to satisfy the HTTP request
+     * if the cache should be used.
+     * <p>
+     * This attribute must be set
+     */
+    public static final String CACHE_KEY = "repose.cache.key";
+
+    /**
+     * Designates whether or not to overwrite the cache data for the
+     * given cache key if cache data was not used to satisfy the request.
+     */
+    public static final String CACHE_FORCE_REFRESH = "repose.cache.refresh";
+
+    public static CachingHttpClientContext adapt(final HttpContext context) {
+        if (context instanceof CachingHttpClientContext) {
+            return (CachingHttpClientContext) context;
+        } else {
+            return new CachingHttpClientContext(context);
+        }
+    }
+
+    public static CachingHttpClientContext create() {
+        return new CachingHttpClientContext();
+    }
+
+    public CachingHttpClientContext(final HttpContext context) {
+        super(context);
+    }
+
+    public CachingHttpClientContext() {
+        super();
+    }
+
+    public Boolean getUseCache() {
```

Comment:
Why big B? Then we have to worry about NPEs all  over the damn place.

Reply:
I went back and forth and this one. I ended up on `Boolean` since the context itself is an `Object` store. Therefore, by using `Boolean`, I can be consistent with other attributes. So yes, you do have to deal with the potential for `null`, but you have to do that for all attributes rather than just some. Using `Boolean` also allows me to represent three distinct states: true, false, or not set. If I had used `boolean`, I would have to return a default value of either `true` or `false` when getting this attribute out of the context. Since the context is just an object store, I preferred putting that determination in the place where we actually use the value (thus giving it meaning).